net/portmapper/igd_test.go

```diff
@@ -164,5 +184,89 @@ func (d *TestIGD) handlePMPQuery(pkt []byte, src netaddr.IPPort) {
 
 func (d *TestIGD) handlePCPQuery(pkt []byte, src netaddr.IPPort) {
 	d.inc(&d.counters.numPCPRecv)
-	// TODO
+	if len(pkt) < 24 {
+		return
+	}
+	op := pkt[1]
+	pktSrcBytes := [16]byte{}
+	copy(pktSrcBytes[:], pkt[8:24])
+	pktSrc := netaddr.IPFrom16(pktSrcBytes)
+	if pktSrc != src.IP() {
+		// TODO this error isn't fatal but should be rejected by server.
+		// Since it's a test it's difficult to get them the same though.
+		d.t.Logf("mismatch of packet source and source IP: got %v, expected %v\n", pktSrc, src.IP())
+	}
+	switch op {
+	case pcpOpAnnounce:
+		d.inc(&d.counters.numPCPDiscoRecv)
+		if !d.doPCP.Get() {
+			return
+		}
+		resp := buildPCPDiscoResponse(pkt)
+		if _, err := d.pxpConn.WriteTo(resp, src.UDPAddr()); err != nil {
+			d.t.Errorf("failed to write disco reply: %v", err)
```

Comment:
oh, there are failures deep in here. Maybe just increment a counter instead and then the caller can fail on it.

I kinda wanted to keep this test stuff independent of the testing package in case we want to reuse it later in, say, integration tests components where there might not be a *testing.T if we're running this in a different process.